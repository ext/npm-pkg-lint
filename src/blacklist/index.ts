import ava from "./ava";
import emacs from "./emacs";
import eslint from "./eslint";
import fimbullinter from "./fimbullinter";
import git from "./git";
import { directory, extension, filename, rcfile } from "./helpers";
import jest from "./jest";
import node from "./node";
import prettier from "./prettier";
import typescript from "./typescript";
import windows from "./windows";

const blacklist: RegExp[] = [
	/* tests in general */
	directory(".nyc_output"),
	directory("coverage"),
	extension(".lcov"),
	extension("[.-]junit.xml"),
	filename(".istanbul.yml"),
	filename("junit.xml"),
	filename("lcov.info"),
	filename(".airtap.yml"),
	rcfile(".c8rc"),
	rcfile(".nycrc"),

	/* tests */
	directory("examples?"),
	directory("perf"),
	directory("tests?"),
	filename("cypress.json"),
	filename("cypress.config.js"),
	filename("cypress.config.cjs"),
	filename("cypress.config.mjs"),
	filename("cypress.config.ts"),
	filename("karma.config.js"),
	filename("mocha.opts"),
	filename("playwright.config.js"),
	filename("playwright.config.cjs"),
	filename("playwright.config.mjs"),
	filename("playwright.config.ts"),
	filename("test.js"),
	filename("testem.json"),
	filename("testem.yml"),
	filename("wallaby.config.js"),
	filename("wallaby.js"),

	/* linting */
	filename(".htmlvalidate.json"),
	filename(".jshintignore"),
	filename(`tslint.json`),
	filename(`tslint.yaml`),
	filename(".jscs.json"),
	rcfile(`.jscsrc`),
	rcfile(`.jshintrc`),

	/* toolchains */
	filename("rollup.config.js"),
	filename("webpack.config.js"),

	/* task runners */
	filename("[gG]runtfile.js"),
	filename("[gG]ulpfile.[jt]s"),
	filename("dangerfile.[jt]s"),
	filename("gulpfile.babel.js"),
	filename("gulpfile.esm.js"),
	filename("Herebyfile.js"),
	filename("Herebyfile.mjs"),

	/* ci */
	/^\.azure-.*$/,
	directory(".circleci"),
	directory(".github"),
	directory(".gitlab"),
	directory(".?zuul.d"),
	filename(".coveralls.yml"),
	filename(".gitlab-ci.yml"),
	filename(".travis.yml"),
	filename(".?zuul.ya?ml"),
	filename("ISSUE_TEMPLATE.md"),
	filename("Jenkinsfile"),
	filename("PULL_REQUEST_TEMPLATE.md"),
	filename("appveyor.yml"),

	/* editor specific files */
	/(^|\/)\.sw[a-p]/ /* vim */,
	directory(".vscode"),
	extension(".sublime-project"),
	extension(".sublime-workspace"),
	filename(".vscodeignore"),
	filename("Session.vim"),
	filename("Sessionx.vim"),
	directory(".idea"),
	filename(".tm_properties") /* textmate */,

	/* OS */
	directory(".AppleDouble"),
	directory(".DS_Store"),
	directory(".LSOverride"),

	/* logs */
	directory("logs"),
	extension(".log"),

	/* runtime */
	directory("pids"),
	extension(".pid"),
	extension(".pid.lock"),
	extension(".seed"),

	/* misc */
	/^[Jj]akefile(\.js)?$/,
	extension(".bak"),
	extension(".swp"),
	extension(".tgz"),
	extension(".tmp"),
	filename(".env"),
	filename(".svgo.yml"),
	filename(".editorconfig"),
	filename(".ext-prepush"),
	filename(".tonic_example.js"),
	filename(".flowconfig"),
	filename(".cardinalrc"),
	filename("bower.json"),
	filename("commitlint.config.js") /* commitlint */,
	filename("composer.json"),
	filename("netlify.toml"),
	filename("release.config.js") /* semantic release */,

	...ava,
	...emacs,
	...eslint,
	...fimbullinter,
	...git,
	...jest,
	...node,
	...prettier,
	...typescript,
	...windows,
];

export default blacklist;
