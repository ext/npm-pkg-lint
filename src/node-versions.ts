interface NodeVersionDescriptor {
	/** If set this version is EOL since given date */
	eol?: string;
}

/* https://github.com/nodejs/Release/blob/main/schedule.json */
export const nodeVersions: Array<[string, NodeVersionDescriptor]> = [
	["0.10.x", { eol: "2016-10-31" }],
	["0.12.x", { eol: "2016-12-31" }],
	["4.x.x", { eol: "2018-04-30" }],
	["5.x.x", { eol: "2016-06-30" }],
	["6.x.x", { eol: "2019-04-30" }],
	["7.x.x", { eol: "2017-06-30" }],
	["8.x.x", { eol: "2019-12-31" }],
	["9.x.x", { eol: "2018-06-30" }],
	["10.x.x", { eol: "2021-04-30" }],
	["11.x.x", { eol: "2019-06-01" }],
	["12.x.x", { eol: "2022-04-30" }],
	["13.x.x", { eol: "2020-06-01" }],
	["14.x.x", { eol: "2023-04-30" }],
	["15.x.x", { eol: "2021-06-01" }],
	["16.x.x", { eol: "2023-09-11" }],
	["17.x.x", { eol: "2022-06-01" }],
	["18.x.x", { eol: "2025-04-30" }],
	["19.x.x", { eol: "2023-06-01" }],
	["20.x.x", {}],
	["21.x.x", { eol: "2024-06-01" }],
	["22.x.x", {}],
];
