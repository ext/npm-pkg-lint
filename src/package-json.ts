import { type PackageJson } from "./types";
import { type Message } from "./message";
import { type Result } from "./result";
import { nonempty, present, typeArray, typeString, ValidationError, validUrl } from "./validators";
import { isDisallowedDependency } from "./rules/disallowed-dependency";
import { exportsTypesOrder } from "./rules/exports-types-order";
import { outdatedEngines } from "./rules/outdated-engines";
import { verifyEngineConstraint } from "./rules/verify-engine-constraint";
import { typesNodeMatchingEngine } from "./rules/types-node-matching-engine";

export interface VerifyPackageJsonOptions {
	allowedDependencies: Set<string>;
	allowTypesDependencies?: boolean;
	ignoreMissingFields?: boolean;
}

type validator = (key: string, value: unknown) => void;

const fields: Record<string, validator[]> = {
	description: [present, typeString, nonempty],
	keywords: [present, typeArray, nonempty],
	homepage: [present, typeString, validUrl],
	bugs: [present, validUrl],
	license: [present, typeString, nonempty],
	author: [present, nonempty],
	repository: [present, validUrl],
};

function verifyFields(pkg: PackageJson, options: VerifyPackageJsonOptions): Message[] {
	const messages: Message[] = [];

	for (const [field, validators] of Object.entries(fields)) {
		try {
			for (const validator of validators) {
				validator(field, pkg[field]);
			}
		} catch (error) {
			if (
				error instanceof ValidationError &&
				error.validator === present.name &&
				options.ignoreMissingFields
			) {
				continue;
			}
			// istanbul ignore else
			if (error instanceof Error) {
				messages.push({
					ruleId: "package-json-fields",
					severity: 2,
					message: error.message,
					line: 1,
					column: 1,
				});
			} else {
				throw error;
			}
		}
	}

	return messages;
}

function getActualDependency(key: string, version: string): string {
	/* handle npm: prefix */
	if (version.startsWith("npm:")) {
		const [name] = version.slice("npm:".length).split("@", 2);
		return name;
	}

	return key;
}

function verifyDependencies(pkg: PackageJson, options: VerifyPackageJsonOptions): Message[] {
	const messages: Message[] = [];

	for (const [key, version] of Object.entries(pkg.dependencies ?? {})) {
		const dependency = getActualDependency(key, version);

		/* skip dependencies explicitly allowed by the user */
		if (options.allowedDependencies.has(dependency)) {
			continue;
		}

		/* skip @types/* if explicitly allowed by user */
		if (options.allowTypesDependencies && dependency.match(/^@types\//)) {
			continue;
		}

		if (isDisallowedDependency(pkg, dependency)) {
			const name = key === dependency ? dependency : `"${key}" ("npm:${dependency}")`;
			messages.push({
				ruleId: "disallowed-dependency",
				severity: 2,
				message: `"${name}" should be a devDependency`,
				line: 1,
				column: 1,
			});
		}
	}

	return messages;
}

export async function verifyPackageJson(
	pkg: PackageJson,
	filePath: string,
	options: VerifyPackageJsonOptions = { allowedDependencies: new Set() }
): Promise<Result[]> {
	const messages: Message[] = [
		...(await verifyEngineConstraint(pkg)),
		...exportsTypesOrder(pkg),
		...verifyFields(pkg, options),
		...verifyDependencies(pkg, options),
		...outdatedEngines(pkg),
		...typesNodeMatchingEngine(pkg),
	];

	if (messages.length === 0) {
		return [];
	}

	return [
		{
			messages,
			filePath,
			errorCount: messages.filter((it) => it.severity === 2).length,
			warningCount: messages.filter((it) => it.severity === 1).length,
			fixableErrorCount: 0,
			fixableWarningCount: 0,
		},
	];
}
