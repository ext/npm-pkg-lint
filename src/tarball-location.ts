import path from "node:path";
import { type PackageJson } from "./types";

function normalize(name: string): string {
	return name.replace("/", "-").replace(/^@/, "");
}

export function tarballLocation(pkg: PackageJson, pkgPath: string): string {
	const name = normalize(pkg.name);
	return path.join(path.dirname(pkgPath), `${name}-${pkg.version}.tgz`);
}
