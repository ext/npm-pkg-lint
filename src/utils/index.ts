export { type NpmInfoError, isNpmInfoError, npmInfo } from "./npm-info";
