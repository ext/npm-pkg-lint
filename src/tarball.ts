import fs from "node:fs";
import * as tar from "tar";
import { type ReadEntry, Parser } from "tar";
import { isBlacklisted } from "./blacklist";
import { type Message } from "./message";
import { type Result } from "./result";
import { type PackageJson, type PackageJsonExports } from "./types";

export interface TarballMeta {
	/** Path to tarball on disk */
	filePath: string;

	/** Path to use in report (default filePath) */
	reportPath?: string | undefined;
}

interface RequiredFile {
	field: string;
	ruleId: string;
	filename: string;
}

const ruleId = "no-missing-main";

function normalize(filename: string): string {
	return filename.replace(/^[^/]+\//, "");
}

export async function getFileList(filename: string): Promise<string[]> {
	const entries: ReadEntry[] = [];
	await tar.list({
		file: filename,
		strict: true,
		onentry: (entry: ReadEntry) => entries.push(entry),
	});
	return entries.map((entry) => {
		const filename = entry.path as unknown as string;
		return normalize(filename);
	});
}

export async function getFileContent(
	tarball: TarballMeta,
	filenames: string[],
): Promise<Record<string, Buffer>> {
	const contents: Record<string, Buffer> = {};

	return new Promise((resolve, reject) => {
		const t = new Parser({
			filter(path): boolean {
				return filenames.includes(normalize(path));
			},
		});
		t.on("entry", (entry: tar.ReadEntry) => {
			const p = normalize(entry.path);
			contents[p] = Buffer.alloc(0);
			entry.on("data", (data: Buffer) => {
				contents[p] = Buffer.concat([contents[p], data]);
			});
			entry.on("error", (error) => {
				/* eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors -- technical debt */
				reject(error);
			});
		});
		t.on("end", () => {
			resolve(contents);
		});
		const rs = fs.createReadStream(tarball.filePath);
		rs.pipe(t as unknown as NodeJS.WritableStream);
	});
}

export function blacklistedFiles(filelist: string[]): string[] {
	return filelist.filter(isBlacklisted);
}

function normalizeRequiredFiles(
	src: string | string[] | Record<string, string | boolean>,
): string[] {
	if (typeof src === "string") {
		return [src];
	} else if (Array.isArray(src)) {
		return src;
	} else {
		return Object.values(src).filter((it) => it !== false) as string[];
	}
}

function* yieldRequiredFiles(
	src: string | string[] | Record<string, string | boolean>,
	template: Pick<RequiredFile, "field" | "ruleId">,
): Generator<RequiredFile> {
	const files = normalizeRequiredFiles(src);
	for (const filename of files) {
		yield { ...template, filename };
	}
}

function* yieldExportedFiles(
	exports: undefined | null | string | PackageJsonExports,
): Generator<RequiredFile> {
	if (!exports) {
		return;
	}

	if (typeof exports === "string") {
		yield {
			field: "exports",
			ruleId: "no-missing-exports",
			filename: exports,
		};
		return;
	}

	for (const [key, value] of Object.entries(exports)) {
		/* ignore wildcards for now */
		if (key.includes("*")) {
			continue;
		}
		yield* yieldExportedFiles(value);
	}
}

function* requiredFiles(pkg: PackageJson): Generator<RequiredFile> {
	yield* yieldExportedFiles(pkg.exports);
	if (pkg.main) {
		yield* yieldRequiredFiles(pkg.main, { field: "main", ruleId });
	}
	if (pkg.browser) {
		yield* yieldRequiredFiles(pkg.browser, { field: "browser", ruleId });
	}
	if (pkg.module) {
		yield* yieldRequiredFiles(pkg.module, { field: "module", ruleId });
	}
	if (pkg["jsnext:main"]) {
		yield* yieldRequiredFiles(pkg["jsnext:main"], {
			field: "jsnext:main",
			ruleId,
		});
	}
	if (pkg.typings) {
		yield* yieldRequiredFiles(pkg.typings, { field: "typings", ruleId: "no-missing-typings" });
	}
	if (pkg.bin) {
		yield* yieldRequiredFiles(pkg.bin, { field: "bin", ruleId: "no-missing-binary" });
	}
	if (pkg.man) {
		yield* yieldRequiredFiles(pkg.man, { field: "man", ruleId: "no-missing-man" });
	}
}

function fileExists(filelist: string[], filename: string): boolean {
	/* strip leading ./ */
	filename = filename.replace(/^\.\//, "");

	/* exact match for filename */
	if (filelist.includes(filename)) {
		return true;
	}

	/* try to append ".js" */
	if (filelist.includes(`${filename}.js`)) {
		return true;
	}

	/* test if it is actually a directory with index.js */
	if (filelist.includes(`${filename.replace(/\/$/, "")}/index.js`)) {
		return true;
	}

	return false;
}

/**
 * @param pkg - Parsed `package.json` data
 * @param tarball - Tarball paths
 */
export async function verifyTarball(pkg: PackageJson, tarball: TarballMeta): Promise<Result[]> {
	const messages: Message[] = [];
	const filelist = await getFileList(tarball.filePath);

	for (const filename of blacklistedFiles(filelist)) {
		messages.push({
			ruleId: "no-disallowed-files",
			severity: 2,
			message: `${filename} is not allowed in tarball`,
			line: 1,
			column: 1,
		});
	}

	for (const requiredFile of requiredFiles(pkg)) {
		if (!fileExists(filelist, requiredFile.filename)) {
			messages.push({
				ruleId: requiredFile.ruleId,
				severity: 2,
				message: `${requiredFile.filename} (pkg.${requiredFile.field}) is not present in tarball`,
				line: 1,
				column: 1,
			});
		}
	}

	if (messages.length === 0) {
		return [];
	}

	return [
		{
			messages,
			filePath: tarball.reportPath ?? tarball.filePath,
			errorCount: messages.filter((it) => it.severity === 2).length,
			warningCount: messages.filter((it) => it.severity === 1).length,
			fixableErrorCount: 0,
			fixableWarningCount: 0,
		},
	];
}
