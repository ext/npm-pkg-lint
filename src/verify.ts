import { type Result } from "./result";
import { type TarballMeta, verifyTarball } from "./tarball";
import { type VerifyPackageJsonOptions, verifyPackageJson } from "./package-json";
import { verifyShebang } from "./shebang";
import { type PackageJson } from "./types";

export type VerifyOptions = VerifyPackageJsonOptions;

export async function verify(
	pkg: PackageJson,
	pkgPath: string,
	tarball: TarballMeta,
	options: VerifyOptions
): Promise<Result[]> {
	return [
		...(await verifyTarball(pkg, tarball)),
		...(await verifyPackageJson(pkg, pkgPath, options)),
		...(await verifyShebang(pkg, tarball)),
	];
}
