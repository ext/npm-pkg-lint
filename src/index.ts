/* eslint-disable no-console, no-process-exit -- this is a cli tool */
/* eslint-disable camelcase -- argparse likes snake_case */
/* eslint-disable security/detect-non-literal-fs-filename -- want to read given filenames */

import { existsSync, createWriteStream, promises as fs } from "fs";
import path from "path";
import { ArgumentParser } from "argparse";
import { findUp } from "find-up";
import tmp from "tmp";
import stylish from "@html-validate/stylish";
import { setupBlacklist } from "./blacklist";
import { verify, VerifyOptions } from "./verify";
import PackageJson from "./types/package-json";
import { tarballLocation } from "./tarball-location";
import { getFileContent, TarballMeta } from "./tarball";

/* eslint-disable-next-line @typescript-eslint/no-var-requires */
const { version } = require("../package.json");

const PACKAGE_JSON = "package.json";

interface ParsedArgs {
	pkgfile: string;
	tarball?: string;
	ignore_missing_fields?: boolean;
	allow_types_dependencies?: boolean;
}

interface GetPackageJsonResults {
	pkg?: PackageJson;
	pkgPath?: string;
}

async function preloadStdin(): Promise<string> {
	return new Promise((resolve, reject) => {
		tmp.file((err, path, fd) => {
			if (err) {
				reject(err);
				return;
			}

			const st = createWriteStream(null, { fd, autoClose: true });
			process.stdin.pipe(st);
			st.on("finish", () => {
				resolve(path);
			});
			st.on("error", (err) => {
				reject(err);
			});
		});
	});
}

async function getPackageJson(
	args: ParsedArgs,
	regenerateReportName: boolean
): Promise<GetPackageJsonResults> {
	/* get from explicit path passed as argument */
	if (args.pkgfile) {
		return {
			pkg: JSON.parse(await fs.readFile(args.pkgfile, "utf-8")),
			pkgPath: args.pkgfile,
		};
	}

	/* extract package.json from explicit tarball location */
	if (args.tarball) {
		const contents = await getFileContent({ filePath: args.tarball }, [PACKAGE_JSON]);
		const pkg = JSON.parse(contents[PACKAGE_JSON].toString("utf-8")); // eslint-disable-line security/detect-object-injection
		return {
			pkg,
			pkgPath: path.join(
				regenerateReportName ? `${pkg.name}-${pkg.version}.tgz` : args.tarball,
				PACKAGE_JSON
			),
		};
	}

	/* try to locate package.json from file structure */
	const pkgPath = await findUp(PACKAGE_JSON);
	if (pkgPath) {
		const relPath = path.relative(process.cwd(), pkgPath);
		return {
			pkg: JSON.parse(await fs.readFile(relPath, "utf-8")),
			pkgPath: relPath,
		};
	}

	return {};
}

async function run(): Promise<void> {
	const parser = new ArgumentParser({
		description: "npm package linter",
	});

	parser.add_argument("-v", "--version", { action: "version", version });
	parser.add_argument("-t", "--tarball", { help: "specify tarball location" });
	parser.add_argument("-p", "--pkgfile", { help: "specify package.json location" });
	parser.add_argument("--allow-types-dependencies", {
		action: "store_true",
		help: "allow dependencies to `@types/*`",
	});
	parser.add_argument("--ignore-missing-fields", {
		action: "store_true",
		help: "ignore errors for missing fields (but still checks for empty and valid)",
	});

	const args: ParsedArgs = parser.parse_args();

	/* this library assumes the file source can be randomly accessed but with
	 * stdin this is not possible so stdin is read into a temporary file which is
	 * used instead */
	let regenerateReportName = false;
	if (args.tarball === "-") {
		args.tarball = await preloadStdin();
		regenerateReportName = true;
	}

	const { pkg, pkgPath } = await getPackageJson(args, regenerateReportName);

	if (!pkg) {
		console.error("Failed to locate package.json and no location was specificed with `--pkgfile'");
		process.exit(1);
	}

	const tarball: TarballMeta = {
		filePath: args.tarball ?? tarballLocation(pkg, pkgPath),
		reportPath: regenerateReportName ? `${pkg.name}-${pkg.version}.tgz` : undefined,
	};
	if (!existsSync(tarball.filePath)) {
		console.error(`"${tarball.filePath}" does not exist, did you forget to run \`npm pack'?`);
		process.exit(1);
	}

	setupBlacklist(pkg.name);

	const options: VerifyOptions = {
		allowTypesDependencies: args.allow_types_dependencies,
		ignoreMissingFields: args.ignore_missing_fields,
	};

	const results = await verify(pkg, pkgPath, tarball, options);
	const output = stylish(results);
	process.stdout.write(output);

	const totalErrors = results.reduce((sum, result) => {
		return sum + result.errorCount;
	}, 0);

	process.exit(totalErrors > 0 ? 1 : 0);
}

run();
