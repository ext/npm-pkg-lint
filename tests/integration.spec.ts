jest.unmock("tar");

import { promises as fs } from "fs";
import * as path from "path";
import { globSync } from "glob";
import { execa } from "execa";
import { verify } from "../src/verify";
import { type PackageJson } from "../src/types";

const ROOT_DIRECTORY = path.resolve(path.join(__dirname, ".."));
const FIXTURE_DIRECTORY = path.join(__dirname, "fixtures");

const fixtures = globSync("*", { cwd: FIXTURE_DIRECTORY });

async function npmPack(pkg: PackageJson, fixture: string): Promise<string> {
	const dir = path.join(FIXTURE_DIRECTORY, fixture);
	const tarball = path.join(dir, `${pkg.name}-${pkg.version}.tgz`);
	await execa("npm", ["pack"], { cwd: dir });
	return path.relative(ROOT_DIRECTORY, tarball);
}

it.each(fixtures)("%s", async (fixture) => {
	expect.assertions(1);
	const dir = path.join(FIXTURE_DIRECTORY, fixture);
	const pkgPath = path.relative(ROOT_DIRECTORY, path.join(dir, "package.json"));
	const pkg: PackageJson = JSON.parse(await fs.readFile(pkgPath, "utf-8"));
	const tarball = { filePath: await npmPack(pkg, fixture) };
	const result = await verify(pkg, pkgPath, tarball, { allowedDependencies: new Set() });
	expect(result).toMatchSnapshot();
});
