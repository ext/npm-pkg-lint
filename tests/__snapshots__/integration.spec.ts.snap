// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`disallowed-files 1`] = `
"ERROR: my-lib.spec.js is not allowed in tarball (no-disallowed-files) at tests/fixtures/disallowed-files/disallowed-files-1.0.0.tgz
"
`;

exports[`empty-fields 1`] = `
"ERROR: "description" must not be empty (package-json-fields) at tests/fixtures/empty-fields/package.json
  2 |   "name": "empty-fields",
  3 |   "version": "1.0.0",
> 4 |   "description": "",
    |                  ^
  5 |   "keywords": [],
  6 |   "homepage": "",
  7 |   "bugs": "",

ERROR: "keywords" must not be empty (package-json-fields) at tests/fixtures/empty-fields/package.json
  3 |   "version": "1.0.0",
  4 |   "description": "",
> 5 |   "keywords": [],
    |               ^
  6 |   "homepage": "",
  7 |   "bugs": "",
  8 |   "repository": {},

ERROR: "homepage" must be a valid url (https only) (package-json-fields) at tests/fixtures/empty-fields/package.json
  4 |   "description": "",
  5 |   "keywords": [],
> 6 |   "homepage": "",
    |               ^
  7 |   "bugs": "",
  8 |   "repository": {},
  9 |   "license": "",

ERROR: "bugs" must be a valid url (https only) (package-json-fields) at tests/fixtures/empty-fields/package.json
   5 |   "keywords": [],
   6 |   "homepage": "",
>  7 |   "bugs": "",
     |           ^
   8 |   "repository": {},
   9 |   "license": "",
  10 |   "author": "",

ERROR: "license" must not be empty (package-json-fields) at tests/fixtures/empty-fields/package.json
   7 |   "bugs": "",
   8 |   "repository": {},
>  9 |   "license": "",
     |              ^
  10 |   "author": "",
  11 |   "main": "",
  12 |   "files": [],

ERROR: "author" must not be empty (package-json-fields) at tests/fixtures/empty-fields/package.json
   8 |   "repository": {},
   9 |   "license": "",
> 10 |   "author": "",
     |             ^
  11 |   "main": "",
  12 |   "files": [],
  13 |   "engines": {

ERROR: "repository" is missing required "type" property (package-json-fields) at tests/fixtures/empty-fields/package.json
   6 |   "homepage": "",
   7 |   "bugs": "",
>  8 |   "repository": {},
     |                 ^
   9 |   "license": "",
  10 |   "author": "",
  11 |   "main": "","
`;

exports[`missing-binary 1`] = `
"ERROR: bin.js (pkg.bin) is not present in tarball (no-missing-binary) at tests/fixtures/missing-binary/missing-binary-1.0.0.tgz
"
`;

exports[`missing-fields 1`] = `
"ERROR: "description" must be set (package-json-fields) at tests/fixtures/missing-fields/package.json
> 1 | {
    | ^
  2 |   "name": "missing-fields",
  3 |   "version": "1.0.0",
  4 |   "files": [],

ERROR: "keywords" must be set (package-json-fields) at tests/fixtures/missing-fields/package.json
> 1 | {
    | ^
  2 |   "name": "missing-fields",
  3 |   "version": "1.0.0",
  4 |   "files": [],

ERROR: "homepage" must be set (package-json-fields) at tests/fixtures/missing-fields/package.json
> 1 | {
    | ^
  2 |   "name": "missing-fields",
  3 |   "version": "1.0.0",
  4 |   "files": [],

ERROR: "bugs" must be set (package-json-fields) at tests/fixtures/missing-fields/package.json
> 1 | {
    | ^
  2 |   "name": "missing-fields",
  3 |   "version": "1.0.0",
  4 |   "files": [],

ERROR: "license" must be set (package-json-fields) at tests/fixtures/missing-fields/package.json
> 1 | {
    | ^
  2 |   "name": "missing-fields",
  3 |   "version": "1.0.0",
  4 |   "files": [],

ERROR: "author" must be set (package-json-fields) at tests/fixtures/missing-fields/package.json
> 1 | {
    | ^
  2 |   "name": "missing-fields",
  3 |   "version": "1.0.0",
  4 |   "files": [],

ERROR: "repository" must be set (package-json-fields) at tests/fixtures/missing-fields/package.json
> 1 | {
    | ^
  2 |   "name": "missing-fields",
  3 |   "version": "1.0.0",
  4 |   "files": [],"
`;

exports[`missing-main 1`] = `
"ERROR: index.js (pkg.main) is not present in tarball (no-missing-main) at tests/fixtures/missing-main/missing-main-1.0.0.tgz
"
`;
